skpro/distributions/qpd.py

```diff
@@ -0,0 +1,245 @@
+        self.H = pd.DataFrame(self.H, index=index)
+        self.B = pd.DataFrame(self.B, index=index)
+        self.n = pd.DataFrame(self.n, index=index)
+        self.theta = pd.DataFrame(self.theta, index=index)
+        self.delta = pd.DataFrame(self.delta, index=index)
+        self.kappa = pd.DataFrame(self.kappa, index=index)
+
+        super().__init__(index=index, columns=columns)
+
+    def mean(self):
+        def cdf_func(x, theta, kappa, delta, n):
+            cdf_arr = self.phi.cdf(
+                1.0
+                / delta
+                * sinh(
+                    arcsinh(1.0 / kappa * log((x - self.l) / theta))
+                    - arcsinh(n * self.c * delta)
+                )
+            )
+            return cdf_arr
+
+        def exp_func(x, theta, kappa, delta, n):
+            # TODO: scipy.integrate will be removed in scipy 1.12.0
+            pdf = derivative(cdf_func, x, dx=1e-6, args=(theta, kappa, delta, n))
+            return x * pdf
+
+        loc = []
+        for i in self.index:
+            theta = self.theta.loc[i, :].to_numpy()
+            kappa = self.kappa.loc[i, :].to_numpy()
+            delta = self.delta.loc[i, :].to_numpy()
+            n = self.n.loc[i, :].to_numpy()
+            # NOTE: integral interval should be checked, -inf to inf will be NaN
+            l, _ = quad(exp_func, args=(theta, kappa, delta, n), a=0.0, b=np.inf)
+            loc.append(l)
+
+        loc_arr = np.array(loc)
+        return pd.DataFrame(loc_arr, index=self.index, columns=self.columns)
+
+    def var(self):
+        def cdf_func(x, theta, kappa, delta, n):
+            cdf_arr = self.phi.cdf(
+                1.0
+                / delta
+                * sinh(
+                    arcsinh(1.0 / kappa * log((x - self.l) / theta))
+                    - arcsinh(n * self.c * delta)
+                )
+            )
+            return cdf_arr
+
+        def var_func(x, mu, theta, kappa, delta, n):
+            # TODO: scipy.integrate will be removed in scipy 1.12.0
+            pdf = derivative(cdf_func, x, dx=1e-6, args=(theta, kappa, delta, n))
+            return ((x - mu) ** 2) * pdf
+
+        mean = self.mean()
+        var = []
+        for i in self.index:
+            theta = self.theta.loc[i, :].to_numpy()
+            kappa = self.kappa.loc[i, :].to_numpy()
+            delta = self.delta.loc[i, :].to_numpy()
+            n = self.n.loc[i, :].to_numpy()
+            mu = mean.loc[i, :].to_numpy()
+            # NOTE: integral interval should be checked, -inf to inf will be NaN
+            l, _ = quad(var_func, args=(mu, theta, kappa, delta, n), a=0.0, b=np.inf)
+            var.append(l)
+
+        var_arr = np.array(var)
+        return pd.DataFrame(var_arr, index=self.index, columns=self.columns)
+
+    def pdf(self, x):
+        # cdf -> pdf calculation because j-qpd's pdf calculation is bit complex
+        def cdf_func(x, theta, kappa, delta, n):
+            cdf_arr = self.phi.cdf(
```

Comment:
Try to avoid the code duplication here.

Reply:
And in other places :).

Reply:
I have avoided some duplication and relocated